Troublesome includes in `matrix_free.h`

#### Description

`hyper.deal/matrix_free/matrix_free.h` has an include with its implementation (`hyper.deal/matrix_free/matrix_free.templates.h`) which causes linking issues if this header is included in more than one translation unit.

#### Observed behavior

Linking fails with "multiple definitions" error (see references).

#### Expected behavior

A successful build.

#### Steps to reproduce

1. Add `#include <hyper.deal/matrix_free.h>` to multiple header files, which have corresponding source files.
2. Compilation is successful, but linking produces "multiple definition" errors for the members of `dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous` (declared and defined at `hyper.deal/matrix_free/vector_partitioner.h`

The following project can be used to test with `cmake -S . -B build/ -DCMAKE_BUILD_TYPE=Debug && cmake --build build/`:

`CMakeLists.txt`:

```cmake
cmake_minimum_required(VERSION 3.12)

find_package(deal.II 9.4.0 HINTS ${deal.II_DIR} $ENV{DEAL_II_DIR} REQUIRED)
DEAL_II_INITIALIZE_CACHED_VARIABLES()

add_executable(hyperdeal_debug "main.cc" "foo.cc" "bar.cc")
target_compile_features(hyperdeal_debug PRIVATE cxx_std_17)
set_target_properties(hyperdeal_debug PROPERTIES CXX_EXTENSIONS OFF)
DEAL_II_SETUP_TARGET(hyperdeal_debug)
target_link_libraries(hyperdeal_debug hyperdeal)
```

`main.cc`:
```c++
#include "bar.h"

int main()
{
  return 0;
}
```

`foo.h`:
```c++
#ifndef FOO_H
#define FOO_H

#include <hyper.deal/matrix_free/matrix_free.h>

#endif
```

`foo.c`:
```c++
#include "foo.h"
```

`bar.h`:
```c++
#ifndef BAR_H
#define BAR_H

#include <hyper.deal/matrix_free/matrix_free.h>
#include "foo.h"

#endif
```

`bar.cc`:
```c++
#include "bar.h"
```

#### Possible solutions

The reason for the issue is that the class implementation is included in the header file, and so gets pasted to each translation unit which needs the declaration.

1. Explicitly instantiate `hyperdeal::MatrixFree` for some common parameter set, so `matrix_free.templates.h` can be moved out of `matrix_free.h`. Pros: better build performance for library users. Cons: need to specify that common parameter set, longer library build times.
2. Inline all class members. Pros: the behavior and build performance will stay the same. Cons: none?
3. Move member implementations to the definition of `hyperdeal::MatrixFree`. Pros: the behavior and build performance will stay the same. Cons: `matrix_free.h` becomes very messy.
4. Do not include `matrix_free.templates.h` in `matrix_free.h`, advise library users to only include `matrix_free.templates.h` only once per project. Pros: common practice in optionally header-only libraries (see e.g. [toml++](https://marzer.github.io/tomlplusplus/group__configuration.html#ga8219bfd7e75587516385b6022b28abc2)). Cons: easy to make a mistake and get a load of seemingly unrelated errors.

#### References

Linking error log:
```
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_start(unsigned int, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_finish(dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_start(dealii::VectorOperation::values, unsigned int, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_finish(dealii::VectorOperation::values, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_start(unsigned int, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_finish(dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_start(dealii::VectorOperation::values, unsigned int, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_finish(dealii::VectorOperation::values, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::sync(unsigned int) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::reinit(std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<std::pair<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> > >, std::allocator<std::pair<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> > > > >, ompi_communicator_t*, ompi_communicator_t*, bool)'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::get_maps() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::get_maps_ghost() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::memory_consumption() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/foo.cc.o: multiple definition of 'hyperdeal::internal::combine_global_cell_info(ompi_communicator_t*, ompi_communicator_t*, hyperdeal::internal::GlobalCellInfo const&, hyperdeal::internal::GlobalCellInfo const&, hyperdeal::internal::GlobalCellInfo&, int, int)'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_start(unsigned int, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_finish(dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_start(dealii::VectorOperation::values, unsigned int, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_finish(dealii::VectorOperation::values, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<double const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<double const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<double, dealii::MemorySpace::Host> const&, dealii::ArrayView<double const, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_start(unsigned int, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::export_to_ghosted_array_finish(dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_start(dealii::VectorOperation::values, unsigned int, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::import_from_ghosted_array_finish(dealii::VectorOperation::values, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, std::vector<dealii::ArrayView<float const, dealii::MemorySpace::Host>, std::allocator<dealii::ArrayView<float const, dealii::MemorySpace::Host> > > const&, dealii::ArrayView<float, dealii::MemorySpace::Host> const&, dealii::ArrayView<float const, dealii::MemorySpace::Host> const&, std::vector<ompi_request_t*, std::allocator<ompi_request_t*> >&) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::sync(unsigned int) const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::reinit(std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<std::pair<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> > >, std::allocator<std::pair<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> > > > >, ompi_communicator_t*, ompi_communicator_t*, bool)'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::get_maps() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::get_maps_ghost() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'dealii::internal::MatrixFreeFunctions::VectorDataExchange::Contiguous::memory_consumption() const'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: error: CMakeFiles/hyperdeal_debug.dir/bar.cc.o: multiple definition of 'hyperdeal::internal::combine_global_cell_info(ompi_communicator_t*, ompi_communicator_t*, hyperdeal::internal::GlobalCellInfo const&, hyperdeal::internal::GlobalCellInfo const&, hyperdeal::internal::GlobalCellInfo&, int, int)'
/nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld.gold: CMakeFiles/hyperdeal_debug.dir/main.cc.o: previous definition here
```

--- a/include/hyper.deal/matrix_free/matrix_free.templates.h
+++ b/include/hyper.deal/matrix_free/matrix_free.templates.h
@@ -567,7 +567,7 @@ namespace hyperdeal
       std::vector<unsigned int> n2;
     };
 
-    void
+    inline void
     combine_global_cell_info(const MPI_Comm        comm_x,
                              const MPI_Comm        comm_v,
                              const GlobalCellInfo &info_x,

--- a/include/hyper.deal/matrix_free/vector_partitioner.h
+++ b/include/hyper.deal/matrix_free/vector_partitioner.h
@@ -74,7 +74,7 @@ namespace internal
          * Initialize partitioner with a list of locally owned cells and
          * a list of ghost faces (cell and face no).
          */
-        void
+        inline void
         reinit(const std::vector<dealii::types::global_dof_index> local_cells,
                const std::vector<std::pair<dealii::types::global_dof_index,
                                            std::vector<unsigned int>>>
@@ -134,7 +134,7 @@ namespace internal
         /**
          * Start to export to ghost array.
          */
-        void
+        inline void
         export_to_ghosted_array_start(
           const unsigned int                     communication_channel,
           const dealii::ArrayView<const double> &locally_owned_array,
@@ -146,7 +146,7 @@ namespace internal
         /**
          * Finish to export to ghost array.
          */
-        void
+        inline void
         export_to_ghosted_array_finish(
           const dealii::ArrayView<const double> &locally_owned_array,
           const std::vector<dealii::ArrayView<const double>> &shared_arrays,
@@ -156,7 +156,7 @@ namespace internal
         /**
          * Start to import from ghost array.
          */
-        void
+        inline void
         import_from_ghosted_array_start(
           const dealii::VectorOperation::values  vector_operation,
           const unsigned int                     communication_channel,
@@ -169,7 +169,7 @@ namespace internal
         /**
          * Finish to import from ghost array.
          */
-        void
+        inline void
         import_from_ghosted_array_finish(
           const dealii::VectorOperation::values vector_operation,
           const dealii::ArrayView<double> &     locally_owned_storage,
@@ -181,7 +181,7 @@ namespace internal
         /**
          * Start to export to ghost array.
          */
-        void
+        inline void
         export_to_ghosted_array_start(
           const unsigned int                    communication_channel,
           const dealii::ArrayView<const float> &locally_owned_array,
@@ -193,7 +193,7 @@ namespace internal
         /**
          * Finish to export to ghost array.
          */
-        void
+        inline void
         export_to_ghosted_array_finish(
           const dealii::ArrayView<const float> &locally_owned_array,
           const std::vector<dealii::ArrayView<const float>> &shared_arrays,
@@ -203,7 +203,7 @@ namespace internal
         /**
          * Start to import from ghost array.
          */
-        void
+        inline void
         import_from_ghosted_array_start(
           const dealii::VectorOperation::values vector_operation,
           const unsigned int                    communication_channel,
@@ -216,7 +216,7 @@ namespace internal
         /**
          * Finish to import from ghost array.
          */
-        void
+        inline void
         import_from_ghosted_array_finish(
           const dealii::VectorOperation::values vector_operation,
           const dealii::ArrayView<float> &      locally_owned_storage,
@@ -307,16 +307,17 @@ namespace internal
         /**
          * Return position of shared cell: cell -> (owner, offset)
          */
-        const std::map<dealii::types::global_dof_index,
-                       std::pair<unsigned int, unsigned int>> &
+        inline const std::map<dealii::types::global_dof_index,
+                              std::pair<unsigned int, unsigned int>> &
         get_maps() const;
 
 
         /**
          * Return position of ghost face: (cell, no) -> (owner, offset)
          */
-        const std::map<std::pair<dealii::types::global_dof_index, unsigned int>,
-                       std::pair<unsigned int, unsigned int>> &
+        inline const std::map<
+          std::pair<dealii::types::global_dof_index, unsigned int>,
+          std::pair<unsigned int, unsigned int>> &
         get_maps_ghost() const;
 
         /**
@@ -324,13 +325,13 @@ namespace internal
          *
          * @note: Only counts the buffers [TODO].
          */
-        std::size_t
+        inline std::size_t
         memory_consumption() const;
 
         /**
          * Synchronize.
          */
-        void
+        inline void
         sync(const unsigned int tag = 0) const;
 
       private:
@@ -399,7 +400,7 @@ namespace internal
       };
 
 
-      void
+      inline void
       Contiguous::export_to_ghosted_array_start(
         const unsigned int                     communication_channel,
         const dealii::ArrayView<const double> &locally_owned_array,
@@ -418,7 +419,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::export_to_ghosted_array_finish(
         const dealii::ArrayView<const double> &             locally_owned_array,
         const std::vector<dealii::ArrayView<const double>> &shared_arrays,
@@ -433,7 +434,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::import_from_ghosted_array_start(
         const dealii::VectorOperation::values  vector_operation,
         const unsigned int                     communication_channel,
@@ -454,7 +455,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::import_from_ghosted_array_finish(
         const dealii::VectorOperation::values vector_operation,
         const dealii::ArrayView<double> &     locally_owned_storage,
@@ -473,7 +474,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::export_to_ghosted_array_start(
         const unsigned int                    communication_channel,
         const dealii::ArrayView<const float> &locally_owned_array,
@@ -492,7 +493,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::export_to_ghosted_array_finish(
         const dealii::ArrayView<const float> &             locally_owned_array,
         const std::vector<dealii::ArrayView<const float>> &shared_arrays,
@@ -507,7 +508,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::import_from_ghosted_array_start(
         const dealii::VectorOperation::values vector_operation,
         const unsigned int                    communication_channel,
@@ -528,7 +529,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::import_from_ghosted_array_finish(
         const dealii::VectorOperation::values vector_operation,
         const dealii::ArrayView<float> &      locally_owned_storage,
@@ -547,7 +548,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::sync(const unsigned int tag) const
       {
         std::vector<MPI_Request> req(sm_targets.size() + sm_sources.size());
@@ -666,7 +667,7 @@ namespace internal
 
 
 
-      void
+      inline void
       Contiguous::reinit(
         const std::vector<dealii::types::global_dof_index> local_cells,
         const std::vector<
@@ -1791,8 +1792,8 @@ namespace internal
 
 
 
-      const std::map<dealii::types::global_dof_index,
-                     std::pair<unsigned int, unsigned int>> &
+      inline const std::map<dealii::types::global_dof_index,
+                            std::pair<unsigned int, unsigned int>> &
       Contiguous::get_maps() const
       {
         return maps;
@@ -1800,8 +1801,9 @@ namespace internal
 
 
 
-      const std::map<std::pair<dealii::types::global_dof_index, unsigned int>,
-                     std::pair<unsigned int, unsigned int>> &
+      inline const std::map<
+        std::pair<dealii::types::global_dof_index, unsigned int>,
+        std::pair<unsigned int, unsigned int>> &
       Contiguous::get_maps_ghost() const
       {
         return maps_ghost;
@@ -1809,7 +1811,7 @@ namespace internal
 
 
 
-      std::size_t
+      inline std::size_t
       Contiguous::memory_consumption() const
       {
         // [TODO] not counting maps and maps_ghost

